xds/internal/xdsclient/callback_serializer.go

```diff
@@ -0,0 +1,88 @@
+/*
+ *
+ * Copyright 2022 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package xdsclient
+
+import (
+	"google.golang.org/grpc/internal/buffer"
+	"google.golang.org/grpc/internal/grpcsync"
+)
+
+// CallbackSerializer provides a mechanism to schedule callbacks in a
+// synchronized manner. It provides a FIFO guarantee on the order of execution
+// of scheduled callbacks.
+//
+// New callbacks can be scheduled by invoking the Schedule() method. To cleanup
+// resources used by the serializer and to discard any queued callbacks, the
+// Close() method needs to be called.
+//
+// This type is safe for concurrent access.
+type CallbackSerializer struct {
+	closed    *grpcsync.Event
+	done      *grpcsync.Event
+	callbacks *buffer.Unbounded
+}
+
+// NewCallbackSerializer returns a new CallbackSerializer instance.
+func NewCallbackSerializer() *CallbackSerializer {
+	t := &CallbackSerializer{
+		closed:    grpcsync.NewEvent(),
+		done:      grpcsync.NewEvent(),
+		callbacks: buffer.NewUnbounded(),
+	}
+	go t.run()
+	return t
+}
+
+// Close shuts down the CallbackSerializer. It is guaranteed that no callbacks
+// will be scheduled once this function returns. If a callback is currently
+// being executed, this functions blocks until execution of that callback
+// finishes before returning.
+func (t *CallbackSerializer) Close() {
+	if t.closed.HasFired() {
+		return
+	}
+	t.closed.Fire()
+	<-t.done.Done()
```

Comment:
It seems like nothing ever clears out `t.callbacks`.  Maybe this is okay if `t` is dereferenced, it should get garbage collected automatically?

Should `buffer.Unbounded` just have its own `Close` method that we leverage instead of needing another `closed` event here?

Reply:
I think it would be nice to have a `Close()` or a `Drain()` method on `buffer.Unbounded`. After the recent memory leak that we have, I'm leaning towards being defensive and cleaning up as much as we can.

I could introduce such a function, and change the behavior of existing functions as follows:
- Put(): after close, would be a nop
- Load(): after close, would be a nop
- Get(): after close, would return a channel on which it is guaranteed that nothing would be sent

But there are a lot of users of `buffer.Unbounded` currently, and I can do this cleanup in a follow up PR, if you agree to the semantics change on the other methods as listed above.

Let me know what you think.

Reply:
Something like that seems plausible.  Let's discuss more offline.